plugins/inputs/nfsclient/nfsclient.go

```diff
@@ -0,0 +1,421 @@
+	tags := map[string]string{"mountpoint": mountpoint, "serverexport": export}
+	nline := convert(line)
+	first := strings.Replace(line[0], ":", "", 1)
+
+	var fields = make(map[string]interface{})
+
+	if fullstat && first == "events" && len(nline) >= len(eventsFields) {
+		for i, t := range eventsFields {
+			fields[t] = nline[i]
+		}
+		acc.AddFields("nfs_events", fields, tags)
+	} else if fullstat && first == "bytes" && len(nline) >= len(bytesFields) {
+		for i, t := range bytesFields {
+			fields[t] = nline[i]
+		}
+		acc.AddFields("nfs_bytes", fields, tags)
+	} else if fullstat && first == "xprt" && len(line) > 1 {
+		switch line[1] {
+		case "tcp":
+			if len(nline)+2 >= len(xprttcpFields) {
+				for i, t := range xprttcpFields {
+					fields[t] = nline[i+2]
+				}
+				acc.AddFields("nfs_xprt_tcp", fields, tags)
+			}
+		case "udp":
+			if len(nline)+2 >= len(xprtudpFields) {
+				for i, t := range xprtudpFields {
+					fields[t] = nline[i+2]
+				}
+				acc.AddFields("nfs_xprt_udp", fields, tags)
+			}
+		}
+	} else if fullstat {
+		if version == "3" {
+			if nfs3Ops[first] && (len(nline) <= len(nfsopFields)) {
+				for i, t := range nline {
+					item := fmt.Sprintf("%s_%s", first, nfsopFields[i])
+					fields[item] = t
+				}
+				acc.AddFields("nfs_ops", fields, tags)
+			}
+		} else if version == "4" {
+			if nfs4Ops[first] && (len(nline) <= len(nfsopFields)) {
+				for i, t := range nline {
+					item := fmt.Sprintf("%s_%s", first, nfsopFields[i])
+					fields[item] = t
+				}
+				acc.AddFields("nfs_ops", fields, tags)
+			}
+		}
+	} else {
+		if first == "READ" {
+			fields["read_ops"] = nline[0]
+			fields["read_retrans"] = (nline[1] - nline[0])
+			fields["read_bytes"] = (nline[3] + nline[4])
+			fields["read_rtt"] = nline[6]
+			fields["read_exe"] = nline[7]
+			acc.AddFields("nfsstat_read", fields, tags)
+		} else if first == "WRITE" {
+			fields["write_ops"] = nline[0]
+			fields["write_retrans"] = (nline[1] - nline[0])
+			fields["write_bytes"] = (nline[3] + nline[4])
+			fields["write_rtt"] = nline[6]
+			fields["write_exe"] = nline[7]
+			acc.AddFields("nfsstat_write", fields, tags)
+		}
+	}
+
+	return nil
+}
+
+func (n *NFSClient) processText(scanner *bufio.Scanner, acc telegraf.Accumulator) error {
+	var device string
+	var version string
+	var export string
+	var skip bool
+	var nfs3Ops map[string]bool
+	var nfs4Ops map[string]bool
+	nfs3Ops = make(map[string]bool)
+	nfs4Ops = make(map[string]bool)
+
+	if len(n.IncludeOperations) == 0 {
+		for _, Op := range nfs3Fields {
+			nfs3Ops[Op] = true
+		}
+		for _, Op := range nfs4Fields {
+			nfs4Ops[Op] = true
+		}
+	} else {
+		for _, Op := range n.IncludeOperations {
+			nfs3Ops[Op] = true
+		}
+		for _, Op := range n.IncludeOperations {
+			nfs4Ops[Op] = true
+		}
+	}
+
+	if len(n.ExcludeOperations) > 0 {
+		for _, Op := range n.ExcludeOperations {
+			if nfs3Ops[Op] {
+				delete(nfs3Ops, Op)
+			}
+			if nfs4Ops[Op] {
+				delete(nfs4Ops, Op)
+			}
+		}
+	}
```

Comment:
Can we please move this to an initialization phase like to the `Init()` function? Otherwise you are constructing the map in _EVERY_ `Gather()` cycle.

Reply:
Actually, moving this to an init function means that moving the globals into parseStat() and processText() doesn't make sense any more, so that has to get refactored as well.

I agree its somewhat inefficient.  However, I'd like to get this correct first, and optimize second.  I'm testing this on some hosts with over 1,000 mount points, and it usually takes less than 10ms to parse the entire file.  The 1,000 mounts is why I want to added the include/exclude lists in the first place.

Reply:
It's not about optimization but rather about readability. For code in the init you directly know it's only executed once and required to setup the thing. In gather you only see stuff required to run the collection. For me that's just easier to understand what your intention is. Remember I'm not the author of the code nor do I know the code as deeply as you do... :-)

Reply:
Still open.